### DISCUSSION TITLE
Create `readonly` method

### DISCUSSION THREAD

**Discussion by igorkamyshev** [sergeysova/patronum#162] [Ideas]
# Abstract

In some cases, I should forbid store modification by external modules.

For example, when I write `userService`, I want to encapsulate all modification in this service and provide something like `readonly` store for the rest of the application.

Example:

```ts
// user_service.ts
const $userName = createStore('')
  .on(userNamedChanged, (_, newName) => newName)
  .on(fecthUserNameFx.doneData, (_, fetchedName) => fetchedName)

const userService = { $uerName: readonly($userName) }

export default userService
```

```ts
// other_file_in_the_project.ts

import userService from './user_service'

// It's the rest of the application 🌚
```

So, here I want to forbid `.on` calls on `userService.$userName` and usage `userService.$userName` as a target in any operators.

# Motivation

Encapsulation of write operations is a good way to structure any program. So, I think, we need a way to forbid writes to stores in some cases.

**Comment by sergeysova** [comment]
Implicit: https://share.effector.dev/TBtJRKG3
Explicit: https://share.effector.dev/j3DMfGDf

By @yumauri

**Reply by yumauri** [reply]
Those snippets above ↑ are just an idea and possible direction of thought, do not take them directly in you code!

**Reply by yumauri** [reply]
Also, a bit simpler approach — to check not _data_, but _stack parent node_: https://share.effector.dev/Sj5XSbo5

**Comment by yumauri** [comment]
_Lyrical digression:_

Following my example snippets, one thought came to my mind:

For now, effector’s graph hasn’t any information about edges. All effector’s graph’s edges are just JavaScript references.

What if! We could add some meta-data to edges? Like _color_. And inside node’s pipe we could get meta-data of the edge, by which one we’ve got data to process?

Let’s say, by default, all edges has _no color_. We add one edge, paint it in _black_, and we can filter data — allow only data, which came by this _black_ edge. All API, used after on the node, will add _colorless_ edges (by default) — and we will just ignore data from those new edges.

Same idea, as in my snippets, but embedded inside effector’s core.

Maybe this edges’ meta-data could be used elsewhere, to solve other cases.

**Comment by AlexandrHoroshih** [comment]
Looks like, since the deprecation of `on` on derived store, simple `$store.map(v => v)` would be enough to create readonly version of any store?

UPD:
Effector doesnt warn or errors if mapped store used as target in `sample` :(
https://share.effector.dev/V8YVMxXi

**Reply by sergeysova** [reply]
@zerobias What about to deprecate use ReadonlyStore as target?

**Reply by zerobias** [reply]
Seems reasonable, would you like to create an issue in effector repo for that case?

**Reply by AlexandrHoroshih** [reply]
Issue for the case:
https://github.com/effector/effector/issues/563
